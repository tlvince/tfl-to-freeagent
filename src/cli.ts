#!/usr/bin/env node

import { convert, Params } from '.'

let input = ''
process.stdin.on('data', data => {
  input += data
})

process.stdin.once('end', () => {
  const [, basename, claimentParm, ...claimant] = process.argv

  if (claimentParm !== '--claimant') {
    console.error(`${basename} --claimant <name>`)
    process.exit(1)
  }

  const params: Params = {
    claimant_name: claimant.join(' '), // eslint-disable-line @typescript-eslint/camelcase
  }

  convert(input, params)
    .then(res => process.stdout.write(res))
    .catch(error => {
      console.error(error)
      process.exit(1)
    })
})
