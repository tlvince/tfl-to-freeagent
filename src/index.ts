import neatCsv, { Row } from 'neat-csv' // eslint-disable-line import/default

const freeAgentHeaders = [
  'claimant_name',
  'category',
  'DD/MM/YYYY',
  'currency',
  'type',
  'gross_value',
  'description',
  'receipt_reference',
  'sales_tax_rate',
  'sales_tax_value',
  'ec_status',
  'native_gross_value',
  'native_sales_tax_value',
  'project_client',
  'project_name',
  'rebill_by',
  'rebill_factor_amount',
  'stock_item_name',
  'stock_quantity',
  'asset_life',
]

const tflHeaders = ['Date', 'Daily Charge (GBP)']

const freeAgentToTfl = {
  'DD/MM/YYYY': 'Date',
  gross_value: 'Daily Charge (GBP)', // eslint-disable-line @typescript-eslint/camelcase
}

const isValid = (row: Row): boolean =>
  tflHeaders.every((tflHeader: string): string => row[tflHeader])

const inverseAmount = (str: string) => {
  if (str === '0') return str
  if (str.startsWith('-')) return str.slice(1)
  return '-' + str
}

const transform = (row: Row) => ({
  ...row,
  'Daily Charge (GBP)': inverseAmount(row['Daily Charge (GBP)']),
})

export interface Params {
  claimant_name: string
  category?: string
  currency?: string
  type?: string
  description?: string
  sales_tax_rate?: string
  project_client?: string
  project_name?: string
}

const defaultParams = {
  category: 'Travel',
  currency: 'GBP',
  type: 'Payment',
  description: 'TfL travel',
  sales_tax_rate: 'Auto', // eslint-disable-line @typescript-eslint/camelcase
}

export const convert = async (
  csvLines: string,
  rawParams: Params
): Promise<string> => {
  const params = { ...defaultParams, ...rawParams }

  const rows = await neatCsv(csvLines)
  return rows
    .filter(isValid)
    .map(transform)
    .reduce((acc, parsedRow: Row) => {
      const row = freeAgentHeaders
        .map(header => {
          if (params[header]) {
            return params[header]
          }

          const tflHeader = freeAgentToTfl[header]
          if (tflHeader) {
            return parsedRow[tflHeader]
          }

          return ''
        })
        .join(',')
      return `${acc}${row}\n`
    }, '')
}
