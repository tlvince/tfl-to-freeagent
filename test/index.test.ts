import { convert, Params } from '../src'
import { promises as fsP } from 'fs'

interface SetupResponse {
  csvFixture: string
}

const setup = async (): Promise<SetupResponse> => {
  const buffer = await fsP.readFile('./test/fixtures/tfl.csv')
  const csvFixture = buffer.toString()
  return {
    csvFixture,
  }
}

describe('convert', () => {
  it('should handle a well-formed CSV', async () => {
    const { csvFixture } = await setup()
    /* eslint-disable @typescript-eslint/camelcase */
    const params: Params = {
      claimant_name: 'Foo',
      project_client: 'Fez Consulting Ltd',
      project_name: 'Business Strategy',
    }
    /* eslint-enable @typescript-eslint/camelcase */
    const converted = await convert(csvFixture, params)
    expect(converted).toMatchSnapshot()
  })
})
