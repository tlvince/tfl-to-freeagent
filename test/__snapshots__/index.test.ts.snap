// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convert should handle a well-formed CSV 1`] = `
"Foo,Travel,08/10/2018,GBP,Payment,6.80,TfL travel,,Auto,,,,,Fez Consulting Ltd,Business Strategy,,,,,
"
`;
